src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,439 @@
+"""
+function itercycles(dg::DiGraph)
+    sccs = strongly_connected_components(dg)
+    for scc in sccs
+        while length(scc) > 1
+            wdg, vmap = induced_subgraph(dg, scc)
+            #startnode = 1
+            shift!(scc)
+            visitor = JohnsonVisitor(wdg)
+            circuit(1, wdg, visitor, vmap)
+        end
+    end
+end
+
+"""
+```countcycles(dg::DiGraph, ceiling = 10^6)```
+
+Count the number of cycles in a directed graph, using 
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+The ceiling is here to avoir memory overload if there are a lot of cycles in the graph. 
+Default value is 10^6, but it can be higher or lower. You can use the function 
+```maxcycles{T}(dg::DiGraph, byscc::Bool = true)``` to get an idea of the 
+theoretical maximum number or cycles.
+
+# Arguments
+* `dg`: the directed graph we are interested in
+* `ceiling = 10^6`: the number of cycles after which the search stops
+
+# Returns
+* `i`: the number of cycles if below the ceiling, the ceiling otherwise
+"""
+function countcycles(dg::DiGraph, ceiling = 10^6)
+    t = Task(() -> itercycles(dg))
+    i = -1 
+    while (t.state == :runnable) & (i < ceiling)
+        consume(t)
+        i += 1
+    end
+    return i
+end
+
+"""
+```getcycles(dg::DiGraph, ceiling = 10^6)```
+
+Search all cycles of the given directed graph, using 
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007), 
+up to the ceiling (avoid memory overload).
+
+If the graph is small, the ceiling will not bite and 
+``simplecycles(dg::DiGraph)`` is more efficient. it avoids the overhead of the 
+counting and testing if the ceiling is reached.
+
+To get an idea of the possible number of cycles, using function 
+```maxcycles{T}(dg::DiGraph, byscc::Bool = true)``` on the directed graph. 
+
+# Arguments:
+* `dg`: the directed graph to explore
+* `ceiling = 10^6`: the number of cycles after which the search stops
+
+# Returns:
+* `cycles`: all the cycles of the directed graph.
+"""
+function getcycles(dg::DiGraph, ceiling = 10^6)
+    t = Task(() -> itercycles(dg))
+    i = -1
+    cycles = Vector{Vector{Int}}()
+    while (t.state == :runnable) & (i < ceiling)
+        cycle = consume(t)
+        i += 1
+		if !(cycle == nothing)
+			push!(cycles, cycle)
+		end
+    end
+    return cycles
+end
+
+"""
+```getcycleslength(dg::DiGraph, ceiling = 10^6)```
+
+Search all cycles of the given directed graph, using 
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007), 
+and return their length.
+
+To get an idea of the possible number of cycles, using function 
+```maxcycles(dg::DiGraph, byscc::Bool = true)``` on the directed graph. 
+
+# Arguments:
+* `dg`: the directed graph to explore
+
+# Returns:
+* `cyclelengths`: the lengths of all cycles, the index in the array is the length
+* `ncycles`: the number of cycles in the directed graph, up to the ceiling
+"""
+function getcycleslength(dg::DiGraph, ceiling = 10^6)
```

Comment:
This function could just be `cyclessizes(dg::DiGraph, ceiling  = 10^6)` or something similar.